spring-cloud-azure-appconfiguration-config/src/main/java/com/microsoft/azure/spring/cloud/config/AzureConfigBootstrapConfiguration.java

```diff
@@ -40,80 +33,68 @@
     }
 
-    @Bean
-    public AzureTokenCredentials tokenCredentials(AzureCloudConfigProperties properties) {
-        if (StringUtils.hasText(System.getenv(ENV_MSI_ENDPOINT))
-                && StringUtils.hasText(System.getenv(ENV_MSI_SECRET))) {
-            return new AppServiceMSICredentials(AzureEnvironment.AZURE);
-        }
-
-        AzureManagedIdentityProperties msiProps = properties.getManagedIdentity();
-        MSICredentials credentials = new MSICredentials();
-        if (msiProps != null && msiProps.getClientId() != null) {
-            credentials.withClientId(msiProps.getClientId());
-        } else if (msiProps != null && msiProps.getObjectId() != null) {
-            credentials.withObjectId(msiProps.getObjectId());
-        }
-
-        return credentials;
-    }
-
     @Bean
     public CloseableHttpClient closeableHttpClient() {
         return HttpClients.createDefault();
     }
 
     @Bean
     public AzureConfigPropertySourceLocator sourceLocator(AzureCloudConfigProperties properties,
-            AppConfigProviderProperties appProperties, ClientStore clients) {
-        return new AzureConfigPropertySourceLocator(properties, appProperties, clients);
+            AppConfigProviderProperties appProperties, ClientStore clients, ApplicationContext context) {
+        TokenCredentialProvider tokenCredentialProvider = null;
+        try {
+            tokenCredentialProvider = context.getBean(TokenCredentialProvider.class);
+        } catch (NoUniqueBeanDefinitionException e) {
+            LOGGER.error("Failed to find unique TokenCredentialProvider Bean for authentication.", e);
+            if (properties.isFailFast()) {
+                throw e;
+            }
+        } catch (NoSuchBeanDefinitionException e) {
+            LOGGER.info("No TokenCredentialProvider found.");
+        }
+        return new AzureConfigPropertySourceLocator(properties, appProperties, clients, tokenCredentialProvider);
     }
 
     @Bean
     public ClientStore buildClientStores(AzureCloudConfigProperties properties,
-            AppConfigProviderProperties appProperties, ConnectionStringPool pool) {
-        return new ClientStore(properties, appProperties, pool);
-    }
-
-    @PostConstruct
-    public void collectTelemetry() {
-        TelemetryCollector.getInstance().addService(TELEMETRY_SERVICE);
+            AppConfigProviderProperties appProperties, ConnectionPool pool, ApplicationContext context) {
+        TokenCredentialProvider tokenCredentialProvider = null;
+        try {
+            tokenCredentialProvider = context.getBean(TokenCredentialProvider.class);
+        } catch (NoUniqueBeanDefinitionException e) {
+            LOGGER.error("Failed to find unique TokenCredentialProvider Bean for authentication.", e);
+            if (properties.isFailFast()) {
+                throw e;
+            }
+        } catch (NoSuchBeanDefinitionException e) {
+            LOGGER.info("No TokenCredentialProvider found.");
```

Comment:
Wondering if we need to make sure that ClientStore always has a TokenCredentialProvider here?

Reply:
They don't need to specify a TokenCredentialProvider, an example is a connection-string being specified.